game/groundforces/ground_force_group.py

```diff
@@ -0,0 +1,163 @@
+import logging
+import random
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import ClassVar, TYPE_CHECKING, Any, Iterator, Optional
+
+import yaml
+
+from game import db
+from game.data.groups import GroupRole, GroupTask
+from game.dcs.groundunittype import GroundUnitType
+from game.dcs.shipunittype import ShipUnitType
+from game.dcs.unittype import UnitType
+from game.point_with_heading import PointWithHeading
+from game.groundforces.template import GroundObjectTemplate
+
+if TYPE_CHECKING:
+    from game import Game
+    from game.factions.faction import Faction
+    from game.theater import TheaterGroundObject, ControlPoint
+
+
+@dataclass
+class GroundForceGroup:
+    name: str
+    ground_units: list[GroundUnitType]
+    ship_units: list[ShipUnitType]
+    statics: list[str]
+    role: GroupRole
+    tasks: list[GroupTask] = field(default_factory=list)
+    template_names: list[str] = field(default_factory=list)
+
+    _by_name: ClassVar[dict[str, GroundForceGroup]] = {}
+    _by_role: ClassVar[dict[GroupRole, list[GroundForceGroup]]] = {}
+    _loaded: bool = False
+    _templates: list[GroundObjectTemplate] = field(default_factory=list)
+
+    def __str__(self) -> str:
+        return self.name
+
+    def update_from_unit_group(self, unit_group: GroundForceGroup) -> None:
+        # Update tasking and templates
+        self.tasks.extend([task for task in unit_group.tasks if task not in self.tasks])
+        self._templates.extend(
+            [
+                template
+                for template in unit_group.templates
+                if template not in self.templates
+            ]
+        )
+
+    @property
+    def templates(self) -> list[GroundObjectTemplate]:
+        return self._templates
+
+    def add_template(self, template: GroundObjectTemplate, faction: Faction) -> None:
+        try:
+            self._templates.append(template.for_faction(faction))
+        except StopIteration:
+            logging.warning(f"Template {template.name} not usable by {faction.name}")
+
+    def load_templates(self, faction: Faction) -> None:
+        """Load all possible templates for this group"""
+        for template_name in self.template_names:
+            for template in db.TEMPLATES.by_name(template_name):
+                self.add_template(template, faction)
+
+        if not self._templates:
+            # Find all matching templates if no preferred set or available
+            for template in list(db.TEMPLATES.by_tasks(self.tasks)):
+                if any(self.has_unit_type(unit) for unit in template.units):
+                    self.add_template(template, faction)
+
+    def set_templates(self, templates: list[GroundObjectTemplate]) -> None:
+        self._templates = templates
+
+    def has_unit_type(self, unit_type: UnitType[Any]) -> bool:
+        return unit_type in self.ground_units or unit_type in self.ship_units
+
+    @property
+    def unit_types(self) -> Iterator[str]:
```

Comment:
It would require an implementation of StaticType before this can be migrated to a complex type. It took me many hours to add the ship types already. Currently i would accept the Str downside more than spending more hours on adding the static type. This can be done later anytime. I think i already changed dozens of -str- to complex class already with this rework. It's a bit exhausting.